### PR TITLE
WebGL2 support

### PR DESCRIPTION
**Pull request by pdaehne**
This commit adds WebGL2 support to webgl-debug.js

### CODE REVIEW

File: src/debug/webgl-debug.js [hunk 1]
```diff
@@ -92,31 +92,31 @@ var glValidEnumContexts = {
...
-    7: { 0:true, 4:true, 5:true }
-  },
+  //'texImage2D': {
+  //   9: { 0:true, 2:true, 6:true, 7:true },
+  //   6: { 0:true, 2:true, 3:true, 4:true }
+  //},
+  //'texSubImage2D': {
+  //  9: { 0:true, 6:true, 7:true },
+  //  7: { 0:true, 4:true, 5:true }
+  //},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Let's not leave commented-out code in here.

Does webgl-debug.js still work with WebGL 1.0 contexts after your changes? Or do these changes make it specific to WebGL 2.0? We should keep it working with WebGL 1.0.

Comment (kenrussell) [reply src/debug/webgl-debug.js]: Concretely: I suggest adding a comment here like "texImage2D and texSubImage2D are defined below with WebGL 2.0 changes", and removing the commented-out code.

File: src/debug/webgl-debug.js [hunk 2]
```diff
@@ -92,31 +92,31 @@ var glValidEnumContexts = {
...
   'copyTexImage2D': {8: { 0:true, 2:true }},
   'copyTexSubImage2D': {8: { 0:true }},
   'generateMipmap': {1: { 0:true }},
-  'compressedTexImage2D': {7: { 0: true, 2:true }},
-  'compressedTexSubImage2D': {8: { 0: true, 6:true }},
+  //'compressedTexImage2D': {7: { 0: true, 2:true }},
+  //'compressedTexSubImage2D': {8: { 0: true, 6:true }},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Same here -- add a comment and refer to the entries below.

File: src/debug/webgl-debug.js [hunk 3]
```diff
@@ -92,31 +92,31 @@ var glValidEnumContexts = {
...
 
   // Buffer objects
 
   'bindBuffer': {2: { 0:true }},
-  'bufferData': {3: { 0:true, 2:true }},
-  'bufferSubData': {3: { 0:true }},
+  //'bufferData': {3: { 0:true, 2:true }},
+  //'bufferSubData': {3: { 0:true }},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Same here

File: src/debug/webgl-debug.js [hunk 4]
```diff
@@ -92,31 +92,31 @@ var glValidEnumContexts = {
...
   'getBufferParameter': {2: { 0:true, 1:true }},
 
   // Renderbuffers and framebuffers
 
   'pixelStorei': {2: { 0:true, 1:true }},
-  'readPixels': {7: { 4:true, 5:true }},
+  //'readPixels': {7: { 4:true, 5:true }},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Here too

File: src/debug/webgl-debug.js [hunk 5]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
...
+
+  'getInternalformatParameter': {3: { 0:true, 1:true, 2:true }},
+  'renderbufferStorageMultisample': {5: { 0:true, 2:true }},
+
+  // WebGL 2 Texture objects
+
+  'texStorage2D': {5: { 0:true, 2:true }},
+  'texStorage3D': {6: { 0:true, 2:true }},
+  'texImage2D': {
+    9: { 0:true, 2:true, 6:true, 7:true },
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Could you please add a comment here indicating that the first two texImage2D entries are for WebGL 1.0, and the last is for WebGL 2.0?

File: src/debug/webgl-debug.js [hunk 6]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
...
+    6: { 0:true, 2:true, 3:true, 4:true },
+    10: { 0:true, 2:true, 6:true, 7:true }
+  },
+  'texImage3D': {
+    10: { 0:true, 2:true, 7:true, 8:true },
+    11: { 0:true, 2:true, 7:true, 8:true }
+  },
+  'texSubImage2D': {
+    9: { 0:true, 6:true, 7:true },
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Same here.

File: src/debug/webgl-debug.js [hunk 7]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
...
+    7: { 0:true, 4:true, 5:true },
+    10: { 0:true, 6:true, 7:true }
+  },
+  'texSubImage3D': {
+    11: { 0:true, 8:true, 9:true },
+    12: { 0:true, 8:true, 9:true }
+  },
+  'copyTexSubImage3D': {9: { 0:true }},
+  'compressedTexImage2D': {
+    7: { 0: true, 2:true },
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Similar here.

File: src/debug/webgl-debug.js [hunk 8]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
...
+    8: { 0: true, 2:true },
+    9: { 0: true, 2:true }
+  },
+  'compressedTexImage3D': {
+    8: { 0: true, 2:true },
+    9: { 0: true, 2:true },
+    10: { 0: true, 2:true }
+  },
+  'compressedTexSubImage2D': {
+    8: { 0: true, 6:true },
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Here too.

File: src/debug/webgl-debug.js [hunk 9]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
...
+
+  // WebGL 2 Writing to the drawing buffer
+
+  'drawArraysInstanced': {4: { 0:true }},
+  'drawElementsInstanced': {5: { 0:true, 2:true }},
+  'drawRangeElements': {6: { 0:true, 4:true }},
+
+  // WebGL 2 Reading back pixels
+
+  'readPixels': {7: { 4:true, 5:true }, 8: { 4:true, 5:true }},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Same here

File: src/debug/webgl-debug.js [hunk 10]
```diff
@@ -152,7 +152,126 @@ var glValidEnumContexts = {
 
   // EXT_blend_minmax extension
 
-  'blendEquationEXT': {1: { 0:true }}
+  'blendEquationEXT': {1: { 0:true }},
+
+  // WebGL 2 Buffer objects
+
+  'bufferData': {3: { 0:true, 2:true }, 4: { 0:true, 2:true }, 5: { 0:true, 2:true }},
```
Comment (kenrussell) [inline src/debug/webgl-debug.js]: Could you add comments indicating which entries for bufferData and bufferSubData are for WebGL 1.0 vs. 2.0?

**Review by kenrussell — CHANGES_REQUESTED** [review]
Thanks Patrick, this is a nice upgrade. A couple of questions and requests for a couple of cleanups.

**Comment by pdaehne** [comment]
Hi Ken, made the changes you requested. Yes, webgl-debug.js still works with WebGL 1.

**Review by kenrussell — APPROVED** [review]
Thank you very much Patrick for this nice contribution and your follow-ups. Many developers will appreciate it. Please consider announcing the availability of your fixes on the public_webgl mailing list.

**Comment by pdaehne** [comment]
Thank you Ken for merging my changes!

**Pull request merged**